### ISSUE TITLE
linter 1.32.0 plan

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#3848]
# 1.32.0

- update `avoid_types_as_parameter_names` to handle type variables
- update `avoid_positional_boolean_parameters` to handle typedefs
- improve `unnecessary_parenthesis` support for property accesses and method invocations 
- update `avoid_redundant_argument_values` to check parameters of redirecting constructors
- performance improvements for `prefer_const_literals_to_create_immutables`
- update `use_build_context_synchronously` to check context properties
- fix false positive for `avoid_private_typedef_functions` with generalized type aliases

---

## release testing

**Try:** https://dart-review.googlesource.com/c/sdk/+/275044

### downstream fixes

- [x] fix new `avoid_redundant_argument_values` diagnostics in `flutter` 
    - (https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8797403260884127729/+/u/analyze_flutter_flutter/stdout)
- [x] fix new `unnecessary_parenthesis` diagnostics in the SDK (https://dart-review.googlesource.com/c/sdk/+/275044)
- [x] fix new `unnecessary_parenthesis` diagnostics in `flutter` 
   - [x] https://github.com/flutter/flutter/pull/117081
   - [x] fix false positive with nullable property accesses (🍻 @asashour -- #3902)
   - [x] fix false positive with a postfix `!` (🎉 @asashour -- #3904)
- [x] fix new `unnecessary_parenthesis` diagnostics in `flutter plugins`
  - [x]  https://github.com/flutter/plugins/pull/6841
- [ ]  fix new `unnecessary_parenthesis` diagnostics in `flutter engine` 
  - [x] https://github.com/flutter/engine/pull/38291

---

/fyi @srawlins

**Comment by asashour** [comment]
The link is private, I guess it should better be the public https://dart-review.googlesource.com/c/sdk/+/269747

(there were other places where the the dart-review SDK link didn't contain `googlesource`, but this probably seems important).

**Comment by pq** [comment]
Thanks for catching this!

**Comment by pq** [comment]
Reviving this as 1.31.0 has landed and will (soon 🤞) be cherry-picked to beta.

I'll update the issue description with a changelog.  There's a lot queued up so I expect this could be a pretty fun one to land (but good to do before we start on some plumbing for removals, etc (#3880)).  🥳 

/fyi @srawlins @bwilkerson @oprypin

**Comment by pq** [comment]
Here's an interesting `unnecessary_parenthesis` diagnostic:

```dart
^ (pingSuccessRegex?.pattern).hashCode
```

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart#L553


(and also 559).

See:

https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8794947421050841633/+/u/analyze_flutter_flutter/stdout

/cc @asashour /fyi @srawlins

**Comment by pq** [comment]
Looking at the Flutter, another false positive?

```dart
   dartEntrypointArgs: ((json['dartEntrypointArgs'] as List<dynamic>?)?.cast<String>())!,
```

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/device.dart#L1010

/fyi @asashour @srawlins

**Comment by pq** [comment]
Flutter plugin diagnostic fixes in flight: https://github.com/flutter/plugins/pull/6841

**Comment by srawlins** [comment]
Lol at some point code is just too hard to read.
